doc/contributing/traitsInDepth.md

```diff
@@ -0,0 +1,45 @@
+# Traits in Depth
+
+The Trait system is Terria's way of managing a priority order for state. For example, we might have a default opacity property from a config file that should be overriden by a user provided value. This article will go through how Traits are made, how the priority system works and how updates happen in the UI.
+
+## Making a Trait
+![Making a trait](./img/making_a_trait.png)
+
+For base traits such as `BaseMapsTraits`, we extend ModelTraits. However when composing multiple traits to form a new trait, we use the `mixTraits` function. For example, `ArcGisTerrainCatalogItemTraits` is composed out of `UrlTraits, MappableTraits, CatalogMemberTraits`.
+
+When creating a new property within a Trait, we use [decorators](https://babeljs.io/docs/en/babel-plugin-proposal-decorators) to ensure that the field takes part in the Trait system.
+
+
+## Stratum order model
+![Stratum order model](./img/stratum.png)
+Stratums are how Terria determines which value a Trait should resolve to. There are 5 categories of Stratum:
+
+1. Defaults (lowest priority)
+2. Underride
+3. Definition
+4. Override
+5. User (highest priority)
+
+### Default Stratum
+This is the lowest priority stratum best thought of as a sensible fallback value.
+
+### Underride Stratum
+When loading data from external sources, we might want to set values at a lower priority. For example, if a field exists both at a group level and at an individual member level. We might want to set the value to underride at the group level so that the value can be overriden at the individual member level.
+
+### Definition Stratum
+This is the default priority level we should use for values loaded from configuration whether it be from a config file or from an asynchronous call to an external API.
+
+### Override Stratum
+This is the priority level we use when we want to override a configuration value that isn't a result of a user interaction.
+
+### User Stratum
+This is the highest priority stratum. We use this when a user explicitly sets some value and we want that value to override the values loaded through configuration.
+
+## What happens at runtime
+![Runtime](./img/runtime.png)
```

Comment:
nit: in StratumOrder the ordering topToBottom means highest priority (user) to lowest (defaults). It'd be nice, but not essential, for the diagram to match and order these top to bottom from highest to lowest priority.

And if you make the image again it'd be nice for the accessing code to look more like normal Trait code (just `item.opacity`, or `model.opacity`)

Reply:
Actually have I got that wrong? Is topToBottom actually deafults to user? I cannot read `sort` functions:
https://github.com/TerriaJS/terriajs/blob/65c3956256db46b5a2e2d7a0d7f9284515b84915/lib/Models/Definition/StratumOrder.ts#L89-L121